reactor-core/src/test/java/reactor/core/scheduler/InstantPeriodicWorkerTaskTest.java

```diff
@@ -0,0 +1,300 @@
+        } finally {
+            exec.shutdownNow();
+            Schedulers.resetOnHandleError();
+        }
+    }
+
+    @Test
+    public void disposeOnCurrentThread() {
+        ExecutorService exec = Executors.newSingleThreadExecutor();
+        Disposable.Composite composit = Disposables.composite();
+
+        try {
+            InstantPeriodicWorkerTask task = new InstantPeriodicWorkerTask(errorRunnable, exec, composit);
+
+            task.thread = Thread.currentThread();
+
+            task.dispose();
+
+            FutureTask<Void> f1 = new FutureTask<Void>(emptyRunnable, null);
+            task.setFirst(f1);
+
+            assertThat(f1.isCancelled()).isTrue();
+
+            FutureTask<Void> f2 = new FutureTask<Void>(emptyRunnable, null);
+            task.setRest(f2);
+
+            assertThat(f2.isCancelled()).isTrue();
+        } finally {
+            exec.shutdownNow();
+            Schedulers.resetOnHandleError();
+        }
+    }
+
+    @Test
+    public void firstCancelRace() {
+        ExecutorService exec = Executors.newSingleThreadExecutor();
+        Disposable.Composite composit = Disposables.composite();
+
+        try {
+            for (int i = 0; i < 10000; i++) {
+                final InstantPeriodicWorkerTask task = new InstantPeriodicWorkerTask(errorRunnable, exec, composit);
+
+                final FutureTask<Void>
+		                f1 = new FutureTask<Void>(emptyRunnable, null);
+                Runnable r1 = () -> task.setFirst(f1);
+                Runnable r2 = task::dispose;
+
+                race(r1, r2, Schedulers.single());
+
+                assertTrue(f1.isCancelled());
+                assertTrue(task.isDisposed());
+            }
+        } finally {
+            exec.shutdownNow();
+            Schedulers.resetOnHandleError();
+        }
+    }
+
+    @Test
+    public void restCancelRace() {
+        ExecutorService exec = Executors.newSingleThreadExecutor();
+        Disposable.Composite composit = Disposables.composite();
+
+        try {
+            for (int i = 0; i < 10000; i++) {
+                final InstantPeriodicWorkerTask task = new InstantPeriodicWorkerTask(errorRunnable, exec, composit);
+
+                final FutureTask<Void> f1 = new FutureTask<Void>(emptyRunnable, null);
+                Runnable r1 = () -> task.setRest(f1);
+                Runnable r2 = task::dispose;
+
+                race(r1, r2, Schedulers.single());
+
+                assertTrue(f1.isCancelled());
+                assertTrue(task.isDisposed());
+            }
+        } finally {
+            exec.shutdownNow();
+            Schedulers.resetOnHandleError();
+        }
+    }
+
+    private static List<Throwable> prepareErrorHook() {
+        List<Throwable> errorsCollector = new CopyOnWriteArrayList<>();
+        Schedulers.onHandleError((t, e) -> errorsCollector.add(e));
+        return errorsCollector;
+    }
+
+
+    public static void race(final Runnable r1, final Runnable r2, Scheduler s) {
```

Comment:
Can't this be replaced by `RaceTestUtils`?

Reply:
done